src/app/CommandHandler.h

```diff
@@ -50,8 +53,33 @@ class CommandHandler : public Command
 
 private:
     friend class TestCommandInteraction;
+    friend class CommandHandlerHandleDeleter;
     CHIP_ERROR SendCommandResponse();
     CHIP_ERROR ProcessCommandDataElement(CommandDataElement::Parser & aCommandElement) override;
+    static void FinishPendingWork(System::Layer * aLayer, void * aAppState, CHIP_ERROR aError);
+
+    struct CommandHandlerHandleDeleter
+    {
+        void operator()(CommandHandler * apCommandHandler) const;
+    };
+
+    uint32_t mPendingWorkCount = 0;
+
+public:
+    using CommandHandlerAsyncHandle = std::unique_ptr<CommandHandler, CommandHandlerHandleDeleter>;
```

Comment:
It's odd that you're transferring ownership of the Command object in this async case, but you don't do that in the synchronous case. That assumes then that if the object that the Command object got transferred to destructs for some reason, does that mean the Command response never gets sent? What if there are other responses queued up in that response that need to get sent?

That brings up the other peril with this approach - this approach only works for single-command payloads. If the request contained multiple commands, of which more than 1 want to be dispatched asynch, this does not work, since there can only be 1 owner of the object at a time.

I'm not certain why this needs to be this complicated - a simple ref-count in the Handler object that 'holds off' sending responses, that requires symmetric "Increment" and "Decrement" would suffice.

Reply:
The reason for async release is because that the release of the object might happen on a thread other than CHIP mainloop -- for example, in the event loop of some system callback -- I'm not sure what the current thread model is. However -- it should be OK for the caller to encode the response (at least for the case of single command), but the action of sending the response must happen on the CHIP mainloop thread.

Yes, this coulde be simpler if we can gurantee (or document) that the Handler must be released on CHIP mainloop.

Will try to do some modifications to make some assumptions, and make the procedure simpler.

As for the `Increment` and `Decrement`, it takes consideration that there might be various cases in the encoding routine, and the user might lost (or mess up) command sender reference, haveing a handler for RAII would be safer.

Reply:
> The reason for async release is because that the release of the object might happen on a thread other than CHIP mainloop -- for example, in the event loop of some system callback

The latest on the thread model work is that applications manage how threading and synchronization work, and the stack itself, does not.

The assumption always is that the calls into the stack are made safely, with safety provided by the application. So there is no need to schedule work to achieve that.

On the whole, I do think we're doing way too much RAII to work around lazy developers who don't read documentation. I don't think we should make it so 'managed' that objects get auto reclaimed, or messages sent in un-obvious ways because of bad application code. If the application forgot to enqueue a response to an async message, there are other bugs that are going to happen, so auto destructing the object to ensure we have clean shutdowns with zero leaks is not the top-order bit for us to solve IMHO.

Reply:
This all presumes we have good API documentation, which we don't...